chunky/src/java/se/llbit/math/BigPackedOctree.java

```diff
@@ -0,0 +1,773 @@
+
+    // Return dummy Node, will work if only type and data are used, breaks if children are needed
+    return node;
+  }
+
+  @Override
+  public Material getMaterial(int x, int y, int z, BlockPalette palette) {
+    // Building the dummy node is useless here
+    long nodeIndex = getNodeIndex(x, y, z);
+    long value = getAt(nodeIndex);
+    if(value > 0) {
+      return UnknownBlock.UNKNOWN;
+    }
+    return palette.get(typeFromValue(value));
+  }
+
+  @Override
+  public void store(DataOutputStream output) throws IOException {
+    output.writeInt(depth);
+    storeNode(output, 0);
+  }
+
+  @Override
+  public boolean isInside(Vector3 o) {
+    int x = (int) QuickMath.floor(o.x);
+    int y = (int) QuickMath.floor(o.y);
+    int z = (int) QuickMath.floor(o.z);
+
+    int lx = x >>> depth;
+    int ly = y >>> depth;
+    int lz = z >>> depth;
+
+    return lx == 0 && ly == 0 && lz == 0;
+  }
+
+  /**
+   * Moves the ray to the boundary of the octree.
+   * @param ray Ray that will be moved to the boundary of the octree. The origin, distance and normals will be modified.
+   * @return {@code false} if the ray doesn't intersect the octree.
+   */
+  private boolean enterOctree(Ray ray) {
+    double nx, ny, nz;
+    double octree_size = 1 << depth;
+
+    // AABB intersection with the octree boundary
+    double tMin, tMax;
+    double invDirX = 1 / ray.d.x;
+    if (invDirX >= 0) {
+      tMin = -ray.o.x * invDirX;
+      tMax = (octree_size - ray.o.x) * invDirX;
+
+      nx = -1;
+      ny = nz = 0;
+    } else {
+      tMin = (octree_size - ray.o.x) * invDirX;
+      tMax = -ray.o.x * invDirX;
+
+      nx = 1;
+      ny = nz = 0;
+    }
+
+    double tYMin, tYMax;
+    double invDirY = 1 / ray.d.y;
+    if (invDirY >= 0) {
+      tYMin = -ray.o.y * invDirY;
+      tYMax = (octree_size - ray.o.y) * invDirY;
+    } else {
+      tYMin = (octree_size - ray.o.y) * invDirY;
+      tYMax = -ray.o.y * invDirY;
+    }
+
+    if ((tMin > tYMax) || (tYMin > tMax))
+      return false;
+
+    if (tYMin > tMin) {
+      tMin = tYMin;
+
+      ny = -FastMath.signum(ray.d.y);
+      nx = nz = 0;
+    }
+
+    if (tYMax < tMax)
+      tMax = tYMax;
+
+    double tZMin, tZMax;
+    double invDirZ = 1 / ray.d.z;
+    if (invDirZ >= 0) {
+      tZMin = -ray.o.z * invDirZ;
+      tZMax = (octree_size - ray.o.z) * invDirZ;
+    } else {
+      tZMin = (octree_size - ray.o.z) * invDirZ;
+      tZMax = -ray.o.z * invDirZ;
+    }
+
+    if ((tMin > tZMax) || (tZMin > tMax))
+      return false;
+
+    if (tZMin > tMin) {
+      tMin = tZMin;
+
+      nz = -FastMath.signum(ray.d.z);
+      nx = ny = 0;
+    }
+
+    if (tMin < 0)
+      return false;
+
+    ray.o.scaleAdd(tMin, ray.d);
+    ray.n.set(nx, ny, nz);
+    ray.distance += tMin;
+    return true;
+  }
+
+  /**
+   *  {@inheritDoc}
+   */
+  @Override
+  public boolean enterBlock(Scene scene, Ray ray, BlockPalette palette) {
```

Comment:
déjà vu… is this implementation always the same regardless of the octree implementation? If so, we could move it up to the interface as a default implementation.

Reply:
Most of the code is the same and should even be put into `Octree` itself, the only things that changes is the internal representation of a node and how to access children.
I think the best way to do that would be to add some `NodeId` interface that the `Octree` could manipulate and let its implementation do the specific stuff (guess i was too lazy to implement that)

Reply:
So, i've implemented what i said in my last commit. I'm marking this as resolved but have a look at the commit in question